File: src/lib/commands.ts
import { invoke } from "@tauri-apps/api/core";

export type ClipKind = "text" | "rtf" | "html" | "image" | "file_url";

export interface ClipSummary {
  id: string;
  createdAt: string;
  kind: ClipKind;
  textPreview: string;
  payloadRef: string | null;
  isPinned: boolean;
  tags: string[];
}

export interface AppSettings {
  maxItems: number;
  maxPayloadBytes: number;
  trimWhitespaceForTextDedup: boolean;
  useSamplingHash: boolean;
  retentionDays: number;
}

export function searchClips(query: string, limit = 40, offset = 0): Promise<ClipSummary[]> {
  return invoke("search_clips", { query, limit, offset });
}

export function pasteClip(id: string): Promise<void> {
  return invoke("paste_clip", { id });
}

export function deleteClip(id: string): Promise<void> {
  return invoke("delete_clip", { id });
}

export function pinClip(id: string, pinned: boolean): Promise<void> {
  return invoke("pin_clip", { id, pinned });
}

export function hidePanel(): Promise<void> {
  return invoke("hide_panel");
}

export function getSettings(): Promise<AppSettings> {
  return invoke("get_settings");
}

export function saveSettings(settings: AppSettings): Promise<AppSettings> {
  return invoke("save_settings", { settings });
}

export function hasAccessibilityPermission(): Promise<boolean> {
  return invoke("has_accessibility_permission");
}

export function getClipThumbnail(id: string): Promise<string | null> {
  return invoke("get_clip_thumbnail", { id });
}

export interface FilePreview {
  fileType: string;
  extension: string;
  thumbnail: string | null;
  fileCount: number;
  fileName: string;
}

export function getFilePreview(id: string): Promise<FilePreview | null> {
  return invoke("get_file_preview", { id });
}

export function openAccessibilitySettings(): Promise<void> {
  return invoke("open_accessibility_settings");
}

// ── Export / Import ──────────────────────────────────────────────

export function exportToJson(
  ids?: string[] | null,
  kind?: string | null,
  dateFrom?: string | null,
  dateTo?: string | null
): Promise<string> {
  return invoke("export_to_json", {
    ids: ids ?? null,
    kind: kind ?? null,
    dateFrom: dateFrom ?? null,
    dateTo: dateTo ?? null,
  });
}

export function exportToCsv(
  ids?: string[] | null,
  kind?: string | null,
  dateFrom?: string | null,
  dateTo?: string | null
): Promise<string> {
  return invoke("export_to_csv", {
    ids: ids ?? null,
    kind: kind ?? null,
    dateFrom: dateFrom ?? null,
    dateTo: dateTo ?? null,
  });
}

export interface ImportResult {
  added: number;
  skipped: number;
  failed: number;
  versionWarning?: string;
}

export function importFromJson(json: string, mode: string): Promise<ImportResult> {
  return invoke("import_from_json", { json, mode });
}

export function importFromCsv(csv: string, mode: string): Promise<ImportResult> {
  return invoke("import_from_csv", { csv, mode });
}

// ── Tags ─────────────────────────────────────────────────────────

export interface Tag {
  id: number;
  name: string;
  color: string | null;
  createdAt: string;
}

export function createTag(name: string, color?: string | null): Promise<Tag> {
  return invoke("create_tag", { name, color: color ?? null });
}

export function listTags(): Promise<Tag[]> {
  return invoke("list_tags");
}

export function deleteTag(id: number): Promise<void> {
  return invoke("delete_tag", { id });
}

export function addTagToClip(clipId: string, tagId: number): Promise<void> {
  return invoke("add_tag_to_clip", { clipId, tagId });
}

export function removeTagFromClip(clipId: string, tagId: number): Promise<void> {
  return invoke("remove_tag_from_clip", { clipId, tagId });
}

export function getClipTags(clipId: string): Promise<Tag[]> {
  return invoke("get_clip_tags", { clipId });
}

// ── Rules ────────────────────────────────────────────────────────

export interface Rule {
  id: number;
  name: string;
  pattern: string;
  patternType: string;
  action: string;
  actionValue: string | null;
  enabled: boolean;
  priority: number;
  createdAt: string;
}

export function createRule(
  name: string, pattern: string, patternType: string,
  action: string, actionValue?: string | null
): Promise<Rule> {
  return invoke("create_rule", { name, pattern, patternType, action, actionValue: actionValue ?? null });
}

export function listRules(): Promise<Rule[]> {
  return invoke("list_rules");
}

export function updateRule(
  id: number, name: string, pattern: string, patternType: string,
  action: string, actionValue: string | null, enabled: boolean, priority: number
): Promise<Rule> {
  return invoke("update_rule", { id, name, pattern, patternType, action, actionValue, enabled, priority });
}

export function deleteRule(id: number): Promise<void> {
  return invoke("delete_rule", { id });
}

// ── Data Management ──────────────────────────────────────────────

export function deleteByDateRange(from: string, to: string): Promise<number> {
  return invoke("delete_by_date_range", { from, to });
}

export function deleteSelected(ids: string[]): Promise<number> {
  return invoke("delete_selected", { ids });
}

export function deleteByType(kind: string): Promise<number> {
  return invoke("delete_by_type", { kind });
}

export function autoPrune(retentionDays: number): Promise<number> {
  return invoke("auto_prune", { retentionDays });
}

export function countByType(kind: string): Promise<number> {
  return invoke("count_by_type", { kind });
}

export function countByDateRange(from: string, to: string): Promise<number> {
  return invoke("count_by_date_range", { from, to });
}

export function countPrunable(retentionDays: number): Promise<number> {
  return invoke("count_prunable", { retentionDays });
}

export interface TypeBreakdown {
  kind: string;
  count: number;
  bytes: number;
}

export interface AgeBreakdown {
  range: string;
  count: number;
}

export interface DiskUsage {
  totalItems: number;
  totalBytes: number;
  byType: TypeBreakdown[];
  byAge: AgeBreakdown[];
}

export function getDiskUsage(): Promise<DiskUsage> {
  return invoke("get_disk_usage");
}

export function updateTag(id: number, name: string, color?: string | null): Promise<Tag> {
  return invoke("update_tag", { id, name, color: color ?? null });
}

export function applyRulesToClip(clipId: string): Promise<string[]> {
  return invoke("apply_rules_to_clip", { clipId });
}

export function batchApplyRules(): Promise<number> {
  return invoke("batch_apply_rules");
}

File: src-tauri/src/commands.rs
use std::sync::atomic::Ordering;
use std::thread;
use std::time::{Duration, Instant};

use tauri::{AppHandle, Manager};

use crate::error::PasteError;
use crate::history::{
    AppSettings, ClipKind, ClipSummary, DiskUsage, FilePreview, ImportResult, Rule, Tag,
};
use crate::AppState;
use crate::PASTE_IN_PROGRESS;

/// Maximum time to wait for clipboard write to be acknowledged (ms).
const MAX_PASTE_WAIT_MS: u64 = 200;
/// Polling interval when waiting for clipboard changeCount to update (ms).
const PASTE_POLL_STEP_MS: u64 = 15;

#[tauri::command]
pub fn search_clips(
    state: tauri::State<'_, AppState>,
    query: String,
    limit: usize,
    offset: usize,
) -> Result<Vec<ClipSummary>, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.search(&query, limit, offset).map_err(PasteError::Store)
}

/// Wait for the clipboard changeCount to shift, confirming the write completed.
/// Falls back to a minimum sleep if changeCount is unavailable.
fn wait_for_clipboard_write(bridge: &crate::macos_bridge::ClipboardBridge, count_before: i64) {
    let deadline = Instant::now() + Duration::from_millis(MAX_PASTE_WAIT_MS);
    loop {
        let elapsed = Instant::now();
        if elapsed >= deadline {
            break;
        }
        if let Ok(count) = bridge.change_count() {
            if count != count_before {
                return; // write acknowledged
            }
        }
        let remaining = deadline.saturating_duration_since(Instant::now());
        let step = Duration::from_millis(PASTE_POLL_STEP_MS).min(remaining);
        thread::sleep(step);
    }
}

#[tauri::command]
pub fn paste_clip(
    state: tauri::State<'_, AppState>,
    app: AppHandle,
    id: String,
) -> Result<(), PasteError> {
    let clip = {
        let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
        store.get_clip(&id).map_err(PasteError::Store)?
    };

    PASTE_IN_PROGRESS.store(true, Ordering::SeqCst);

    let count_before = state.bridge.change_count().unwrap_or_default();
    state
        .bridge
        .write_clip(&clip)
        .map_err(PasteError::Clipboard)?;
    hide_panel(app.clone())?;

    wait_for_clipboard_write(&state.bridge, count_before);

    let result = state
        .bridge
        .send_paste_keystroke()
        .map_err(PasteError::Clipboard);
    PASTE_IN_PROGRESS.store(false, Ordering::SeqCst);
    result
}

#[tauri::command]
pub fn delete_clip(
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<(), PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.delete_clip(&id).map_err(PasteError::Store)
}

#[tauri::command]
pub fn pin_clip(
    state: tauri::State<'_, AppState>,
    id: String,
    pinned: bool,
) -> Result<(), PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.pin_clip(&id, pinned).map_err(PasteError::Store)
}

#[tauri::command]
pub fn get_settings(
    state: tauri::State<'_, AppState>,
) -> Result<AppSettings, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.get_settings().map_err(PasteError::Store)
}

#[tauri::command]
pub fn save_settings(
    state: tauri::State<'_, AppState>,
    settings: AppSettings,
) -> Result<AppSettings, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.save_settings(settings).map_err(PasteError::Store)
}

#[tauri::command]
pub fn has_accessibility_permission(state: tauri::State<'_, AppState>) -> bool {
    state.bridge.has_accessibility_permission()
}

#[tauri::command]
pub fn open_accessibility_settings() -> Result<(), PasteError> {
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg("x-apple.systempreferences:com.apple.preference.security?Privacy_Accessibility")
            .status()
            .map_err(|error| PasteError::OpenSettings(error.to_string()))?;
    }
    #[cfg(target_os = "windows")]
    {
        // Windows does not require accessibility permission for paste keystrokes
    }
    Ok(())
}

#[tauri::command]
pub fn get_clip_thumbnail(
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<Option<String>, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.get_clip_thumbnail(&id).map_err(PasteError::Store)
}

#[tauri::command]
pub fn get_file_preview(
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<Option<FilePreview>, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.get_file_preview(&id).map_err(PasteError::Store)
}

#[tauri::command]
pub fn hide_panel(app: AppHandle) -> Result<(), PasteError> {
    let Some(window) = app.get_webview_window("main") else {
        return Ok(());
    };
    window.hide().map_err(|error| PasteError::Clipboard(error.to_string()))
}

// ── Export / Import commands ───────────────────────────────────────

#[tauri::command]
pub fn export_to_json(
    state: tauri::State<'_, AppState>,
    ids: Option<Vec<String>>,
    kind: Option<String>,
    date_from: Option<String>,
    date_to: Option<String>,
) -> Result<String, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    let kind = kind.map(|k| ClipKind::from_str(&k));
    store
        .export_to_json(ids, kind, date_from, date_to)
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn export_to_csv(
    state: tauri::State<'_, AppState>,
    ids: Option<Vec<String>>,
    kind: Option<String>,
    date_from: Option<String>,
    date_to: Option<String>,
) -> Result<String, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    let kind = kind.map(|k| ClipKind::from_str(&k));
    store
        .export_to_csv(ids, kind, date_from, date_to)
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn import_from_json(
    state: tauri::State<'_, AppState>,
    json: String,
    mode: String,
) -> Result<ImportResult, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .import_from_json(&json, &mode)
        .map_err(PasteError::Store)
}

// ── Data Management commands ───────────────────────────────────────

#[tauri::command]
pub fn delete_by_date_range(
    state: tauri::State<'_, AppState>,
    from: String,
    to: String,
) -> Result<usize, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .delete_by_date_range(&from, &to)
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn delete_selected(
    state: tauri::State<'_, AppState>,
    ids: Vec<String>,
) -> Result<usize, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.delete_selected(&ids).map_err(PasteError::Store)
}

#[tauri::command]
pub fn delete_by_type(
    state: tauri::State<'_, AppState>,
    kind: String,
) -> Result<usize, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .delete_by_type(ClipKind::from_str(&kind))
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn auto_prune(
    state: tauri::State<'_, AppState>,
    retention_days: usize,
) -> Result<usize, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.auto_prune(retention_days).map_err(PasteError::Store)
}

#[tauri::command]
pub fn count_by_type(
    state: tauri::State<'_, AppState>,
    kind: String,
) -> Result<usize, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .count_by_type(ClipKind::from_str(&kind))
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn count_by_date_range(
    state: tauri::State<'_, AppState>,
    from: String,
    to: String,
) -> Result<usize, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .count_by_date_range(&from, &to)
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn count_prunable(
    state: tauri::State<'_, AppState>,
    retention_days: usize,
) -> Result<usize, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.count_prunable(retention_days).map_err(PasteError::Store)
}

#[tauri::command]
pub fn get_disk_usage(
    state: tauri::State<'_, AppState>,
) -> Result<DiskUsage, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.get_disk_usage().map_err(PasteError::Store)
}

#[tauri::command]
pub fn import_from_csv(
    state: tauri::State<'_, AppState>,
    csv: String,
    mode: String,
) -> Result<ImportResult, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .import_from_csv(&csv, &mode)
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn create_tag(
    state: tauri::State<'_, AppState>,
    name: String,
    color: Option<String>,
) -> Result<Tag, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .create_tag(&name, color.as_deref())
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn list_tags(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<Tag>, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.list_tags().map_err(PasteError::Store)
}

#[tauri::command]
pub fn delete_tag(
    state: tauri::State<'_, AppState>,
    id: i64,
) -> Result<(), PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.delete_tag(id).map_err(PasteError::Store)
}

#[tauri::command]
pub fn add_tag_to_clip(
    state: tauri::State<'_, AppState>,
    clip_id: String,
    tag_id: i64,
) -> Result<(), PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .add_tag_to_clip(&clip_id, tag_id)
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn remove_tag_from_clip(
    state: tauri::State<'_, AppState>,
    clip_id: String,
    tag_id: i64,
) -> Result<(), PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .remove_tag_from_clip(&clip_id, tag_id)
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn get_clip_tags(
    state: tauri::State<'_, AppState>,
    clip_id: String,
) -> Result<Vec<Tag>, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.get_clip_tags(&clip_id).map_err(PasteError::Store)
}

#[tauri::command]
pub fn create_rule(
    state: tauri::State<'_, AppState>,
    name: String,
    pattern: String,
    pattern_type: String,
    action: String,
    action_value: Option<String>,
) -> Result<Rule, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .create_rule(
            &name,
            &pattern,
            &pattern_type,
            &action,
            action_value.as_deref(),
        )
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn list_rules(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<Rule>, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.list_rules().map_err(PasteError::Store)
}

#[tauri::command]
pub fn update_rule(
    state: tauri::State<'_, AppState>,
    id: i64,
    name: String,
    pattern: String,
    pattern_type: String,
    action: String,
    action_value: Option<String>,
    enabled: bool,
    priority: i64,
) -> Result<Rule, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .update_rule(
            id,
            &name,
            &pattern,
            &pattern_type,
            &action,
            action_value.as_deref(),
            enabled,
            priority,
        )
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn delete_rule(
    state: tauri::State<'_, AppState>,
    id: i64,
) -> Result<(), PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store.delete_rule(id).map_err(PasteError::Store)
}

#[tauri::command]
pub fn update_tag(
    state: tauri::State<'_, AppState>,
    id: i64,
    name: String,
    color: Option<String>,
) -> Result<Tag, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    store
        .update_tag(id, &name, color.as_deref())
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn apply_rules_to_clip(
    state: tauri::State<'_, AppState>,
    clip_id: String,
) -> Result<Vec<String>, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    let clip = store.get_clip(&clip_id).map_err(PasteError::Store)?;
    store
        .apply_rules(&clip_id, &clip.text_preview, clip.kind.as_str())
        .map_err(PasteError::Store)
}

#[tauri::command]
pub fn batch_apply_rules(
    state: tauri::State<'_, AppState>,
) -> Result<usize, PasteError> {
    let store = state.store.lock().map_err(|_| PasteError::LockPoisoned)?;
    let clips = store.search("", 10000, 0).map_err(PasteError::Store)?;
    let mut processed = 0;
    for clip in &clips {
        let kind_str = clip.kind.as_str().to_string();
        if let Ok(actions) = store.apply_rules(&clip.id, &clip.text_preview, &kind_str) {
            if !actions.is_empty() {
                processed += 1;
            }
        }
    }
    Ok(processed)
}
